PyPDF2/_cmap.py

```diff
@@ -57,72 +122,104 @@ def parse_encoding(ft: DictionaryObject, space_code: int) -> Tuple[List[str], in
+        .replace(b">>", b"\n}\n")  # some solution to find it back
+    )
+    ll = cm.split(b"<")
+    for i in range(len(ll)):
+        j = ll[i].find(b">")
+        if j >= 0:
+            ll[i] = ll[i][:j].replace(b" ", b"") + b" " + ll[i][j + 1 :]
+    cm = (b" ".join(ll)).replace(b"[", b" [ ").replace(b"]", b" ]\n ")
+
+    for l in cm.split(b"\n"):
+        if l == b"" or l == b" ":
             continue
-        if "beginbfrange" in l:
+        if b"beginbfrange" in l:
             process_rg = True
-        elif "endbfrange" in l:
+        elif b"endbfrange" in l:
             process_rg = False
-        elif "beginbfchar" in l:
+        elif b"beginbfchar" in l:
             process_char = True
-        elif "endbfchar" in l:
+        elif b"endbfchar" in l:
             process_char = False
         elif process_rg:
-            lst = [x for x in l.split(" ") if x]
+            lst = [x for x in l.split(b" ") if x]
             a = int(lst[0], 16)
             b = int(lst[1], 16)
-            if lst[2] == "[":
+            nbi = len(lst[0])
+            map_dict[-1] = nbi // 2
+            fmt = b"%%0%dX" % nbi
+            if lst[2] == b"[":
                 for sq in lst[3:]:
-                    if "]":
+                    if sq == b"]":
                         break
-                    map_dict[a] = unhexlify(sq).decode("utf-16-be")
+                    map_dict[
+                        unhexlify(fmt % a).decode(
+                            "charmap" if map_dict[-1] == 1 else "utf-16-be"
+                        )
+                    ] = unhexlify(sq).decode("utf-16-be")
                     a += 1
                     assert a > b
             else:
                 c = int(lst[2], 16)
-                fmt = b"%%0%dX" % len(lst[2])
+                fmt2 = b"%%0%dX" % len(lst[2])
                 while a <= b:
-                    map_dict[a] = unhexlify(fmt % c).decode("utf-16-be")
+                    map_dict[
+                        unhexlify(fmt % a).decode(
+                            "charmap" if map_dict[-1] == 1 else "utf-16-be"
+                        )
+                    ] = unhexlify(fmt2 % c).decode("utf-16-be")
                     a += 1
                     c += 1
         elif process_char:
-            lst = [x for x in l.split(" ") if x]
-            a = int(lst[0], 16)
-            map_dict[a] = unhexlify("".join(lst[1:])).decode(
-                "utf-16-be"
-            )  # join is here as some cases where the code was split
-
-    # get
+            lst = [x for x in l.split(b" ") if x]
+            map_dict[-1] = len(lst[0]) // 2
+            while len(lst) > 0:
+                map_dict[
+                    unhexlify(lst[0]).decode(
+                        "charmap" if map_dict[-1] == 1 else "utf-16-be"
+                    )
+                ] = unhexlify(lst[1]).decode(
+                    "utf-16-be"
+                )  # join is here as some cases where the code was split
```

Comment:
```suggestion
                hexlified = lst[0]
                try:
                    unhexlified = unhexlify(hexlified)
                except BinasciiError:
                    # Odd-length string
                    break
                key = unhexlified.decode(
                    "charmap" if map_dict[-1] == 1 else "utf-16-be"
                )

                # join is here as some cases where the code was split
                map_dict[key] = unhexlify(lst[1]).decode("utf-16-be")
```

Reply:
can you show me an exemple of issue inhere ?

Reply:
Sure:

```
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1302, in extract_text
    return self._extract_text(self, self.pdf, space_width, PG.CONTENTS)
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1125, in _extract_text
    cmaps[f] = build_char_map(f, space_width, obj)
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_cmap.py", line 21, in build_char_map
    map_dict, space_code = parse_to_unicode(ft, space_code)
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_cmap.py", line 209, in parse_to_unicode
    unhexlify(fmt % a).decode(
binascii.Error: Odd-length string
```

for https://corpora.tika.apache.org/base/docs/govdocs1/999/999097.pdf